src/internet_identity/src/state/temp_keys.rs

```diff
@@ -0,0 +1,130 @@
+    /// cannot be used to authenticate (similarly to how the browser session key pair cannot be used to
+    /// authenticate if the actual delegated WebAuthn device is removed).
+    ///
+    /// In addition, the temporary key is also linked to an anchor so that even if the same device is
+    /// added to multiple anchors, it can only be used in the context of the anchor it was created for.
+    ///
+    /// Since temp keys can only be added during registration, the the max number of temp keys is
+    /// bounded by the registration rate limit.
+    temp_keys: HashMap<DeviceKey, TempKey>,
+
+    /// Deque to efficiently prune expired temp keys
+    expirations: VecDeque<TempKeyExpiration>,
+}
+
+impl TempKeys {
+    pub fn add_temp_key(
+        &mut self,
+        device_key: &DeviceKey,
+        anchor: AnchorNumber,
+        temp_key: Principal,
+    ) {
+        let tmp_key = TempKey {
+            principal: temp_key,
+            anchor,
+            expiration: time() + TEMP_KEY_EXPIRATION_NS,
+        };
+
+        self.expirations.push_back(TempKeyExpiration {
+            device_key: device_key.clone(),
+            expiration: tmp_key.expiration,
+        });
+        self.temp_keys.insert(device_key.clone(), tmp_key);
+    }
+
+    /// Removes the temporary key for the given device if it exists and is linked to the provided anchor.
+    pub fn remove_temp_key(&mut self, anchor: AnchorNumber, device_key: &DeviceKey) {
+        // we can skip the removal from expirations because there it will be removed
+        // during amortized clean-up operations
+        if let Some(temp_key) = self.temp_keys.get(device_key) {
+            if temp_key.anchor == anchor {
+                // Only remove temp key if the anchor matches
+                self.temp_keys.remove(device_key);
+            }
+        }
+    }
+
+    /// Checks that the temporary key is valid for the given device and anchor.
+    ///
+    /// Requires a mutable reference because it does amortized clean-up of expired temp keys.
+    pub fn check_temp_key(
+        &mut self,
+        caller: &Principal,
+        device_key: &DeviceKey,
+        anchor: AnchorNumber,
+    ) -> Result<(), ()> {
+        self.prune_expired_keys();
+
+        let Some(temp_key) = self.temp_keys.get(device_key) else {
+            return Err(());
+        };
+        if temp_key.expiration < time() {
+            return Err(());
+        }
+        if temp_key.anchor != anchor {
+            return Err(());
+        }
+        if &temp_key.principal != caller {
+            return Err(());
+        }
+        Ok(())
+    }
+
+    fn prune_expired_keys(&mut self) {
+        const MAX_TO_PRUNE: usize = 100;
+
+        let now = time();
+        for _ in 0..MAX_TO_PRUNE {
+            let Some(expiration) = self.expirations.front() else {
+                break;
+            };
+
+            if expiration.expiration > now {
+                break;
+            }
```

Comment:
Could add a note here that we know that the expiration objects are sorted because we control the expiration period (and because `time()` is monotonous) (implicit assumption used here)